packages/core/lib/commands/test.js

```diff
@@ -67,81 +63,27 @@ const command = {
       Environment.detect(config).catch(done);
     }
 
-    let ipcDisconnect;
-    let files = [];
+    let ipcDisconnect, files;
 
-    if (options.file) {
-      files = [options.file];
-    } else if (options._.length > 0) {
-      Array.prototype.push.apply(files, options._);
-    }
-
-    try {
-      if (files.length === 0) {
-        const directoryContents = glob.sync(
-          `${config.test_directory}${path.sep}*`
-        );
-        files =
-          directoryContents.filter(item => fs.statSync(item).isFile()) || [];
-      }
-    } catch (error) {
-      return done(error);
-    }
-
-    files = files.filter(function(file) {
-      return file.match(config.test_file_extension_regexp) != null;
-    });
-
-    let temporaryDirectory;
     try {
-      temporaryDirectory = temp.mkdirSync("test-");
+      files = determineTestFilesToRun(options, config);
     } catch (error) {
       return done(error);
     }
 
-    function runCallback() {
-      var args = arguments;
-      // Ensure directory cleanup.
-      done.apply(null, args);
-      if (ipcDisconnect) ipcDisconnect();
-    }
-
-    function run() {
-      // Set a new artifactor; don't rely on the one created by Environments.
-      // TODO: Make the test artifactor configurable.
-      config.artifactor = new Artifactor(temporaryDirectory);
-
-      const testConfig = config.with({
-        test_files: files,
-        contracts_build_directory: temporaryDirectory
-      });
-      Test.run(testConfig)
-        .then(runCallback)
-        .catch(runCallback);
-    }
-
-    const environmentCallback = function() {
-      // Copy all the built files over to a temporary directory, because we
-      // don't want to save any tests artifacts. Only do this if the build directory
-      // exists.
-      try {
-        fs.statSync(config.contracts_build_directory);
-      } catch (_error) {
-        return run();
-      }
-
-      promisifiedCopy(config.contracts_build_directory, temporaryDirectory)
-        .then(() => {
-          config.logger.log("Using network '" + config.network + "'." + OS.EOL);
-
-          run();
-        })
-        .catch(done);
-    };
-
     if (config.networks[config.network]) {
       Environment.detect(config)
-        .then(() => environmentCallback())
+        .then(() => copyArtifactsToTempDir(config))
+        .then(({ config, temporaryDirectory }) => {
+          return prepareConfigAndRunTests({
+            config,
+            files,
+            temporaryDirectory
+          });
+        })
+        .then(numberOfFailures => {
+          done.call(null, numberOfFailures);
```

Comment:
One questionable thing here; I changed `apply` to `call` and did not pass `arguments` which might allow for passing of things other than just the number of failures. I couldn't see anything else that might be passed besides the failures but maybe I am missing something? This way seemed much cleaner and self-descriptive so I favored it.

Reply:
The difference is just in how the arguments are passed, right? I prefer `call` personally, agree that it is cleaner.

Reply:
Hm, the only consideration I'd add here about `call` vs. `apply` is this that `call` is faster, as @MrRefactoring brought to our attention in https://github.com/trufflesuite/truffle/pull/2462

Reply:
I think it is preferable to use call, as it works faster and is more popular among developers.

I think it makes sense to use apply only in one case if you have an array of arguments. This will be faster than using the call and spread operator at the same time.